# Voting Ensemble 

- Its very simple, we train diffrent model with data and take the majority count for classification problem and if its is regrssion we find out tht mean .. 


The baseline model are diffrent models, each are given new query point and they generate thier predection/output and we take majority count . 

And regarding creating variety either 

1. you can train diffrent model with data 
2. you can train the same model but with diffrent dataset so that the diffrence/variety is created 

we are doing for classificaiton problem 

In [2]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 

from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split

data = load_breast_cancer()

X = data.data
y = data.target

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [3]:
X


array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
        1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
        8.902e-02],
       [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
        8.758e-02],
       ...,
       [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
        7.820e-02],
       [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
        1.240e-01],
       [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
        7.039e-02]], shape=(569, 30))

In [4]:
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
       0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0,

Voting are also of 2 types:

1. Hard voting 
- where we jsut take the majority count of the predection and the class with majority count is our answer

2. soft voting :

- Here we take the mean of the probabilits of predection and higher probab wins 

classification problem generally gives the ans as probabilty say model 1 had 0.7 for yes then it would be 0.3 for no 

likewise model 2 had 0.6 for yes and 0.4 for now then 

we calc yes=0.7+0.6/2 = ans1
 no=0.4+0.3/2 =ans2 and ans with more prpobailty is final answer 
 

In [ ]:
from sklearn.ensemble import VotingClassifier
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score


# say for our base model we take 4 models 

lr=LogisticRegression()
dt=DecisionTreeClassifier()
knn=KNeighborsClassifier()


# look at sklarn docs for voting ensemlbe , estimator rewquire list of tuple 
estimators=[('logistic',lr),('tree',dt),('knn',knn)]



voting_ensemble=VotingClassifier(estimators=estimators, voting='hard') # hard is default 
voting_ensemble.fit(X_train, y_train)

y_pred = voting_ensemble.predict(X_test)

print("Hard voting accuracy:", accuracy_score(y_test, y_pred))




soft_voting = VotingClassifier(
    estimators=estimators,
    voting="soft"
)

soft_voting.fit(X_train, y_train)

y_pred = soft_voting.predict(X_test) # classificaiotn so it ueses mode(majority count) to calc final answer

print("Soft voting accuracy:", accuracy_score(y_test, y_pred))


Hard voting accuracy: 0.9385964912280702
Soft voting accuracy: 0.956140350877193


/Library/Frameworks/Python.framework/Versions/3.14/lib/python3.14/site-packages/sklearn/linear_model/_logistic.py:406: ConvergenceWarning: lbfgs failed to converge after 100 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=100).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/Library/Frameworks/Python.framework/Versions/3.14/lib/python3.14/site-packages/sklearn/linear_model/_logistic.py:406: ConvergenceWarning: lbfgs failed to converge after 100 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=100).
You might also want to scale the data as shown in:
   

In [ ]:
# lets crossvalidate the models and see how will it peform 
# we read it genraly should have bigger accuracy then all models , but not always though lts see.

for estimator in estimators:
    corssval=cross_val_score(estimator[1],X,y,cv=5)
    print(estimator[0],np.mean(corssval))

# estimator[1] gets the model object.
# First iteration: ('logistic', lr) -> estimator[1] is lr
# Second iteration: ('tree', dt) -> estimator[1] is dt
# Third iteration: ('knn', knn) -> estimator[1] is knn

# cross_val_score trains and evaluates the model using
# 5-fold cross-validation and returns 5 accuracy scores.
# np.mean() computes the average accuracy across the 5 folds. in corssval also we calc the mean which is our final ans there 

logistic 0.9490451793199813
tree 0.9173420276354605
knn 0.9279459711224964


/Library/Frameworks/Python.framework/Versions/3.14/lib/python3.14/site-packages/sklearn/linear_model/_logistic.py:406: ConvergenceWarning: lbfgs failed to converge after 100 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=100).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/Library/Frameworks/Python.framework/Versions/3.14/lib/python3.14/site-packages/sklearn/linear_model/_logistic.py:406: ConvergenceWarning: lbfgs failed to converge after 100 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=100).
You might also want to scale the data as shown in:
   

so we can see above :
Hard voting accuracy: 0.9385964912280702
Soft voting accuracy: 0.956140350877193. 


then we used corss validation trained model 5 times and validated it and our accuracy or indv model :

logistic 0.9490451793199813
tree 0.9173420276354605
knn 0.9279459711224964


so we can clearly here using voting ensemlve(soft ) accuracy is more than all the individual models ... so we use ensemble if we had to choose an individual model then we woudl choose logistic as it peforms best than others.. 

# at the end, for classificaiton problem, 

- take this as note its imp that 

- For a majority voting classifier, the ensemble often performs better than the
  individual base models, but this is NOT guaranteed.

- If the base models are all better than random guessing (i.e., each has an
  accuracy greater than 50% for a balanced binary classification problem) and
  they make different errors (their errors are not perfectly correlated),
  the voting classifier is likely to achieve a higher accuracy than the
  individual models.

- If a base model performs worse than random guessing (accuracy < 50%),
  including it in the voting ensemble can hurt the ensemble's performance.

- We proved this using probability in class. Refer to those notes for the
  mathematical proof.